### PR TITLE
[SPARK-9828] [PySpark] Mutable values should not be default arguments

### PR DESCRIPTION
**Pull request by MechCoder**

### CODE REVIEW

**Comment by MechCoder** [comment]
@mengxr It might not matter since extra is not mutated internally.

But better safe than sorry.

**Comment by mengxr** [comment]
@MechCoder Do we have other `{}` usage in the implementation?

**Comment by MechCoder** [comment]
I grepped for those and added af ew more.. I think that should be it.

**Comment by MechCoder** [comment]
Ohh, just saw your edited comment. Yes that is indeed a cause of concern

**Comment by MechCoder** [comment]
@mengxr Please look at the last commit.

These variables should be initialized for every instance rather than being created only once during object creation.

**Comment by MechCoder** [comment]
retest this please

**Comment by MechCoder** [comment]
@mengxr tests pass.

File: python/pyspark/ml/wrapper.py [hunk 1]
```diff
@@ -45,10 +45,12 @@ class JavaWrapper(Params):
 
     __metaclass__ = ABCMeta
 
-    #: The wrapped Java companion object. Subclasses should initialize
-    #: it properly. The param values in the Java object should be
-    #: synced with the Python wrapper in fit/transform/evaluate/copy.
-    _java_obj = None
```
Comment (mengxr) [inline python/pyspark/ml/wrapper.py]: I don't think different instances share the same `_java_obj`. Only the default values are shared.

**Comment by mengxr** [comment]
Only one minor issue. Ping @davies for another pass.

File: python/pyspark/ml/wrapper.py (hunk 1 above)
Comment (davies) [inline python/pyspark/ml/wrapper.py]: +1, this change is not necessary.

**Comment by davies** [comment]
It's safe to use mutable object as default, when we know that we will not mutate it. The good part is that we can generate better API doc (see the type and default valule in signature). So I don't see it as a hard rule.

Once we change the default value to None, we should also make sure that the docstring could cover that.

**Comment by MechCoder** [comment]
> It's safe to use mutable object as default, when we know that we will not mutate it.

Indeed but one does not know when someone can mutate it inadvertently inside in future versions (especially when the LOC's are very large)

File: python/pyspark/ml/wrapper.py (hunk 1 above)
Comment (MechCoder) [inline python/pyspark/ml/wrapper.py]: Did you mean this is safe because `_java_obj` is immutable?

Comment (davies) [inline python/pyspark/ml/wrapper.py]: Yes.

Comment (MechCoder) [inline python/pyspark/ml/wrapper.py]: OK. But there was some other reason for changing it. I'll dig deeper and let you know.

**Comment by MechCoder** [comment]
retest this please

**Comment by MechCoder** [comment]
retest this please

**Comment by MechCoder** [comment]
test this please

**Comment by MechCoder** [comment]
@davies Could you please tell me how the docstrings are changed? There should be virtually no change in how the user uses these functions, since if nothing is provided the parameter is assumed to be a list / dict.

**Comment by MechCoder** [comment]
test this please

**Comment by MechCoder** [comment]
The test failures seem unrelated. Are they?

**Comment by rxin** [comment]
@davies does this look good to you?

**Comment by rxin** [comment]
I discussed with @mengxr offline about this. It seems to me it is safer to not use {} at all (i.e. always immutable) in the arguments, even if the docstring is slightly uglier.

**Comment by mengxr** [comment]
@MechCoder I searched `={}` and `=[]` in the codebase and found one more:

~~~
ml/pipeline.py:        __init__(self, stages=[])
ml/pipeline.py:        setParams(self, stages=[])
~~~

Could you update this PR? Thanks!

**Comment by MechCoder** [comment]
Hmm.. those are just docstrings, but I'll still change them.

**Comment by mengxr** [comment]
Merged into master, branch-1.5, and branch-1.4. Thanks @MechCoder!

**Pull request closed without merging**